sql: implement ALTER TABLE LOCALITY REGIONAL BY ROW -> !(REGIONAL BY ROW)

otan opened the pull request:
Migrating from REGIONAL BY ROW is a matter of removing the partition by
column, which is another ALTER PRIMARY KEY style re-index. This is the
technique we use to implement this functionality in this commit.

Release note (sql change): Implemented functionality to ALTER TABLE SET
LOCALITY from REGIONAL BY ROW to REGIONAL BY TABLE or GLOBAL.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60311)
<!-- Reviewable:end -->

arulajmani reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60311)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDjEJ91BUrYOREJCii:-MTDjEJ91BUrYOREJCij:b-t458vs) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/ccl/partitionccl/regional_by_row_test.go#L40)):*
> ```Go
> // TestAlterTableLocalityRegionalByRowError tests an alteration involving
> // REGIONAL BY ROW which gets interrupted.
> func TestAlterTableLocalityRegionalByRowError(t *testing.T) {
> ```

How do you feel about splitting this test into `TestAlterTableLocalityFromRegionalByRow` and `TestTableLocalityToRegionalByRow`? I feel doing this 2-D matching is kinda tricky even now, when things are fresh, so I can only assume be even harder in the future.
___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDkNlb7Laa_x4Pk_Z6:-MTDkNlc-ujtflQAjy3Y:b8wu3j4) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/ccl/partitionccl/regional_by_row_test.go#L64)):*
> ```Go
> 	const showZoneConfigurationSQL = `SHOW ZONE CONFIGURATION FROM TABLE t.test`
> 
> 	// alterState is a struct that contains an action for a base test case to alter towards.
> ```

Did you mean alter from instead of alter towards?
___
*[pkg/sql/alter_primary_key.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDlLu66ubC24NQ7rk2:-MTDlLu66ubC24NQ7rk3:b2jxk11) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/sql/alter_primary_key.go#L251)):*
> ```Go
> 	// isNewPartitionAllBy is set if a new PARTITON ALL BY statement is introduced.
> 	isNewPartitionAllBy := false
> 	// dropPartitionAllBy is set if we should be dropping from PARTITION ALL BY.
> ```

nit: s/from/the/g?
___
*[pkg/sql/alter_table_locality.go, line 284 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDmwyD6IrLxQ0Zdd2c:-MTDmwyD6IrLxQ0Zdd2d:bjjxqi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/sql/alter_table_locality.go#L284)):*
> ```Go
> }
> 
> // alterTableLocalityRegionalByRowChange processes a change for ALTER TABLE ...
> ```

Doesn't this process a change from a REGIONAL BY ROW to either GLOBAL or REGIONAL BY TABLE? In which case, the Alter statement in the comment needs to be updated?
___
*[pkg/sql/alter_table_locality.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDnAgxEKK9n0D98NCu:-MTDnAgxEKK9n0D98NCv:b-t7zb33) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/sql/alter_table_locality.go#L286)):*
> ```Go
> // alterTableLocalityRegionalByRowChange processes a change for ALTER TABLE ...
> // LOCALITY REGIONAL BY ROW, based on the new locality and PK columns.
> func (n *alterTableSetLocalityNode) alterTableLocalityRegionalByRowChange(
> ```

nit: rename to alterTableLocalityFromRegionalByRow or something? As is, it took me until the call site to realize what this function really does. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60311)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @arulajmani)

___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDjEJ91BUrYOREJCii:-MTDr-Fy1IMjkokHKUkQ:b-9o7vqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/ccl/partitionccl/regional_by_row_test.go#L40)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

How do you feel about splitting this test into `TestAlterTableLocalityFromRegionalByRow` and `TestTableLocalityToRegionalByRow`? I feel doing this 2-D matching is kinda tricky even now, when things are fresh, so I can only assume be even harder in the future.
</blockquote></details>

the point of doing it this way was to de-duplicate all the setup code and such. i've fixed the commenting up to be easier to understand though.
___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDkNlb7Laa_x4Pk_Z6:-MTDsA8g49Ml1gCSZGy_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/ccl/partitionccl/regional_by_row_test.go#L64)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Did you mean alter from instead of alter towards?
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_locality.go, line 284 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDmwyD6IrLxQ0Zdd2c:-MTDsdMB2DnVk2Urx5Br:b-cti52z) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/sql/alter_table_locality.go#L284)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Doesn't this process a change from a REGIONAL BY ROW to either GLOBAL or REGIONAL BY TABLE? In which case, the Alter statement in the comment needs to be updated?
</blockquote></details>

it does both -- it is called from alterTableLocalityNonRegionalByRowToRegionalByRow too. comment updated.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60311)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDjEJ91BUrYOREJCii:-MTET-vo8xqA1cakM3Jo:b2npt1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/ccl/partitionccl/regional_by_row_test.go#L40)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

the point of doing it this way was to de-duplicate all the setup code and such. i've fixed the commenting up to be easier to understand though.
</blockquote></details>

That's a fair point -- not sure if Ive stared at this long enough, or the new set of comments is helping, but this does feel a bit better now. Left some suggestions around what changes I would recommend as a reader of this code, but feel free to ignore if you don't agree with me/think they're obvious.
___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTERsj10lgZc19FMACE:-MTERsj10lgZc19FMACF:b4toil4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb477dbdcde9a3a5ee770609e3a6f1f2deb8c1b2/pkg/ccl/partitionccl/regional_by_row_test.go#L39)):*
> ```Go
> 
> // TestAlterTableLocalityRegionalByRowError tests an alteration involving
> // REGIONAL BY ROW which gets interrupted.
> ```

nit: s/interrupted/cancelled. 

Could you add more context here as well, such as we expect the PK to remain the same as it was when the test began and we don't expect there to be a spurious crdb_region column in the table. 
___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTER3YY-nW6jLPN8MKH:-MTER3YY-nW6jLPN8MKI:bv9jeh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb477dbdcde9a3a5ee770609e3a6f1f2deb8c1b2/pkg/ccl/partitionccl/regional_by_row_test.go#L77)):*
> ```Go
> 	nonRegionalByRowAlterStates := []alterState{
> 		{
> 			desc:                  "REGIONAL BY ROW AS cr",
> ```

nit: It might help to prefix these descriptions with "ALTERED TO" or something to that effect. 
___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTETI3oA3b20MJ2f8mR:-MTETI3oA3b20MJ2f8mS:bdd9i2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb477dbdcde9a3a5ee770609e3a6f1f2deb8c1b2/pkg/ccl/partitionccl/regional_by_row_test.go#L113)):*
> ```Go
> 
> 	// testCases contain a base table structure to start off as.
> 	// For each of these test case, we fork a subtest to perform each alterState operation
> ```

nit: Consider rewording to something like:

For each of these test cases, we pick either the alter state corresponding to the base table's locality -- for REGIONAL BY ROW tables we pick `regionalyByRowAlterStates` and for GLOBAL/REGIONAL BY TABLE tables we pick `nonRegionalByRowAlterStates`.
___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTERbNy86IgSC5XrwLr:-MTERbNy86IgSC5XrwLs:b-m492ux) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb477dbdcde9a3a5ee770609e3a6f1f2deb8c1b2/pkg/ccl/partitionccl/regional_by_row_test.go#L143)):*
> ```Go
> 			desc:           "REGIONAL BY ROW",
> 			setupQuery:     `CREATE TABLE t.test (k INT NOT NULL, v INT) LOCALITY REGIONAL BY ROW`,
> 			expectedPKCols: []string{"crdb_region", "rowid"},
> ```

nit: maybe rename this to originalPKCols? I think it helps explain why this is expected after the alter pk fails. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60311)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @arulajmani)

___
*[pkg/ccl/partitionccl/regional_by_row_test.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60311#-MTDjEJ91BUrYOREJCii:-MTEYvdMCrd8UClM_7D9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d08ba8d07f74007e5da3cfd587c07dd5686be1/pkg/ccl/partitionccl/regional_by_row_test.go#L40)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

That's a fair point -- not sure if Ive stared at this long enough, or the new set of comments is helping, but this does feel a bit better now. Left some suggestions around what changes I would recommend as a reader of this code, but feel free to ignore if you don't agree with me/think they're obvious.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
sql: implement ALTER TABLE LOCALITY REGIONAL BY ROW -> !(REGIONAL BY ROW)

Migrating from REGIONAL BY ROW is a matter of removing the partition by
column, which is another ALTER PRIMARY KEY style re-index. This is the
technique we use to implement this functionality in this commit.

Release note (sql change): Implemented functionality to ALTER TABLE SET
LOCALITY from REGIONAL BY ROW to REGIONAL BY TABLE or GLOBAL.

otan commented:
thanks

bors r=arulajmani
